### DISCUSSION TITLE
Batchtools/SLURM/array jobs

### DISCUSSION THREAD

**Discussion by sakrejda** [ropensci/targets#563] [General]
For running large pipelines on a SLURM cluster, it can take a long time to submit a lot of small jobs (but they are *very* convenient for organizing the pipeline).  Array jobs solve this issue in slurm since they can be submitted much faster (at least on the clusters I've worked with) and they are implemented in batchtools but I have tried to make array jobs work with `tar_make_future` and so far no luck.

Has anyone else gotten this combination to work?

My alternative that seems to be fine is the clustermq approach but I sometimes run into weird memory issues on the main node with `clustermq` so I was hoping to have an alternative.

**Comment by wlandau** [comment]
The memory issues in `tar_make_clustermq()` may be easier to solve. Have you tried `tar_option_set(storage = "worker", retrieval = "worker")` so the workers store and retrieve the data?

**Reply by wlandau** [reply]
I am assuming your cluster has a network file system and the workers have the same working directory and mounted storage as the main node. Also, `memory = "transient"` and `garbage_collection = TRUE` are also related.

**Reply by sakrejda** [reply]
That's exactly as far as I've gotten.  One caveat is that I'm not *sure* the clustermq memory issue comes up when I actually run the job on the cluster.  It does come up when I use clustermq locally to run the pipeline on a workstation and it occurred to me that the implementations for those two scenarios might be different that I'm scaling up some tests (with clustermq) on the cluster itself.  I'll update the post if that ends up working out.  

But also it sounds like you're saying that my array job question might just have a negative answer (no straightforward path to using targets/future/batchtools/SLURM with array jobs)

**Reply by wlandau** [reply]
`tar_make_future()` launches one `future::future()` per target, and `future` is mostly based on transient workers. I highly doubt array jobs are possible as persistent workers. If they are, this is completely under the control of `future.batchtools`. But I also suspect array jobs vs single jobs are not the bottleneck here. I highly prefer `tar_make_clustermq()` on my company's SGE cluster due to https://github.com/HenrikBengtsson/future.batchtools/issues/65.

**Comment by stuvet** [comment]
I've got a fairly large pipeline working like this on Slurm. I wanted to use `tar_make_future` to keep costs down as fitting mcmc models on different-sized datasets means some branches within the same target finish in 5min & others take >24hr.

I came up against the same issues as you did - v slow submission. I adjusted some of the network mount settings on my cluster to no effect. Because I'm using scaleable partitions (from 0) I also hit a race condition where sometimes jobs would run after recruiting a new worker & sometimes they'd trigger a 'log file does not exist' error. `tar_make_future` was 100% unusable for me.

Long story short, both problems were solved by these PR [1](https://github.com/mllg/batchtools/pull/276#issue-696300102) & [2](https://github.com/mllg/batchtools/pull/277#issue-696304462). The first bug caused each of my jobs to wait for >65sec after they started before `batchtools` decided they'd started correctly, or they hit the race condition above. After the fix my jobs submit as fast as I'd expect (~2sec for some pretty big jobs) & I don't get expired errors after recruiting a new worker. The `batchtools` PR isn't tested on any setup but mine, so YMMV - use with caution.

I also looked into array jobs to solve some problems, but I can't see any progress since [this](https://github.com/HenrikBengtsson/future.batchtools/issues/23#issuecomment-432353357).

One thing that also really helped with my submission speed was to minimise the size of the working environment & avoid using the `packages` argument in the target spec. All packages are either in tar_options or available to workers via a shared repo (specified in .Rprofile) & required functions are fully qualified when called from a worker (e.g. brms::brm).

Hope that helps some. Good luck.

**Reply by sakrejda-umich** [reply]
Thanks for all that! Really helpful to see someone else tried to pursue the batchtools/array job route as having very small targets can be really convenient and it breaks the targets abstraction to have to organize them around scheduler quirks!  Ultimately I'm having pretty good luck with `clustermq` although I'm not sure if I can track the memory requirements of the (very) heterogenous jobs in my pipeline this way.

**Reply by stuvet** [reply]
@sakrejda-umich I've submitted a more solid PR for the `batchtools` issues I mentioned above. Rare job state codes returned by slurm should be mapped with reasonable default choices after [this](https://github.com/mllg/batchtools/pull/277#issue-696304462).

If you want to try `tar_make_future` again make sure you're on targets 141859fab4492e0566af257d8256157509ad4e5f as the fix there has made my life awesome after a fair bit of pain.

**Reply by sakrejda-umich** [reply]
Fantastic! I really appreciate the follow-up.  I saw your other thread too.  I've been having good luck with `clustermq` *but* since I have to request resources for long-running workers it gives up the fine-grained control that's nice for a heterogeneous pipeline and I'm still going to try this out.  I'll just post here (or your other SLURM issue) if I get it going in a way that works for me.

**Comment by stemangiola** [comment]
Great discussion, at WEHI (Au) we are trying to convert a makeflow workflow (make style SLURM compatible workflow) to R Targets.

However, either `batchtools`/`Targets` combo does not recognise memory overflows, or `clustermq` does not spawn single workers per job.

As our brainstorming is at the preliminary phase, we (2 in the team) would appreciate if anyone would like to have a brief discussion (e.g. via Zoom), to do some back and forth on possible solutions/alternatives, working setups, and perspective, before we embark in this months-long project. 

For academic folks, this project will result in a publication, we would very happy to welcome co-authors on this.

Any availability would be highly appreciated!
Thanks.

Stefano Mangiola

**Reply by wlandau** [reply]
AFAIK there are no plans in `clustermq` to support transient workers, but it would be good to read/post in https://github.com/mllg/batchtools to ask about memory limit issues. 

I am slowly starting a second attempt at https://github.com/wlandau/crew, which if successful will provide another option and a continuum between persistent and transient workers.
